tests/model/cycle_data/test_perfherder_cycling.py

```diff
@@ -595,6 +595,136 @@ def test_stalled_data_removal(
+    test_perf_signature.last_updated = four_months_ago_timestamp
+    test_perf_signature.save()
+
+    push = Push.objects.first()
+
+    perf_data = []
+
+    data_points_count = 2 * nr_months
+    for i in range(0, data_points_count):
+        if i % 2 == 0:
+            timestamp += timedelta(days=30)
+
+        data = PerformanceDatum.objects.create(
+            repository=test_perf_signature.repository,
+            push=push,
+            job=None,
+            signature=test_perf_signature,
+            push_timestamp=timestamp,
+            value=1.0,
+        )
+        perf_data.append(data)
+    call_command('cycle_data', 'from:perfherder')
+
+    assert test_perf_signature in PerformanceSignature.objects.all()
+    for data_point in perf_data:
+        assert data_point in PerformanceDatum.objects.all()
+
+
+def test_big_density_in_historical_data(
+    test_repository,
+    test_perf_signature_2,
+    test_perf_data,
+):
+    nr_months = 8
+    test_repository.name = 'autoland'
+    test_repository.save()
+    four_months_ago_timestamp = datetime.now() - timedelta(days=120)
+    timestamp = datetime.now() - timedelta(days=((nr_months * 30) + 120 + 30))
+    test_perf_signature = test_perf_signature_2
+    test_perf_signature.last_updated = four_months_ago_timestamp
+    test_perf_signature.save()
+
+    push = Push.objects.first()
+
+    perf_data = []
+
+    data_points_count = 2 * nr_months
+
+    # add one data point per month
+    for i in range(0, nr_months):
+        timestamp += timedelta(days=30)
+
+        data = PerformanceDatum.objects.create(
+            repository=test_perf_signature.repository,
+            push=push,
+            job=None,
+            signature=test_perf_signature,
+            push_timestamp=timestamp,
+            value=1.0,
+        )
+        perf_data.append(data)
+
+    # add the remaining data points to the last month which has data
+    for i in range(0, (data_points_count - nr_months)):
+        data = PerformanceDatum.objects.create(
+            repository=test_perf_signature.repository,
+            push=push,
+            job=None,
+            signature=test_perf_signature,
+            push_timestamp=timestamp,
+            value=1.0,
+        )
+        perf_data.append(data)
+
+    call_command('cycle_data', 'from:perfherder')
+
+    assert test_perf_signature in PerformanceSignature.objects.all()
+    for data_point in perf_data:
+        assert data_point in PerformanceDatum.objects.all()
+
+
+def test_non_historical_stalled_data_is_removed(
+    test_repository,
+    test_perf_signature_2,
+    test_perf_data,
+):
+    nr_months = 8
+    test_repository.name = 'autoland'
+    test_repository.save()
+    four_months_ago_timestamp = datetime.now() - timedelta(days=120)
+    timestamp = datetime.now() - timedelta(days=((nr_months * 30) + 120 + 30))
+    test_perf_signature = test_perf_signature_2
+    test_perf_signature.last_updated = four_months_ago_timestamp
```

Comment:
I see we prepare this same signature in all 3 tests. We could consider DRYing it.